core/FrontController.php

```diff
@@ -418,6 +419,10 @@ public function init()
             $this->makeAuthenticator($sessionAuth); // Piwik\Auth must be set to the correct Login plugin
         }
 
+        if (Piwik::isUserIsAnonymous() && Piwik::getModule() !== 'API') {
```

Comment:
@flamisz We'd need to make sure to always execute the browser check when the `Widgetize` API is used. Ideally, for simplicity and to make sure it runs in all cases we would simply always run it when the `Widgetize` module is requested. You may be logged in when viewing the Widgets (either through a session or token).

Regarding the API we would need to add a check like `$skipBrowserCheck = Piwik::getModule() === 'API' && (empty(Piwik::getAction()) || Piwik::getAction() === 'index' || Piwik::getAction() === 'glossary'))`. When you call the API, then basically the `index` action (which is the default action) will be executed. The browser check should not be executed for the `glossary` either (which you can see btw here: https://glossary.matomo.org/). For all other methods in the API controller  (like `listAllAPI`) we would still want to check the browser. 

And there's one more thing :) If the `[General] enable_framed_pages=1` or `[General]enable_framed_settings=1` config is enabled, then we probably also always want to execute the browser check unless it is the API. So in these cases we also want to execute the check when user is not anonymous. This feature is tricky and we don't recommend enabling this feature which is documented in https://matomo.org/faq/troubleshooting/faq_147/ but it basically allows any action to be embedded like if it was Widgetized.  You can find some information about the config here: https://developer.matomo.org/guides/piwiks-ini-configuration . I'm thinking instead of checking the config there is a more simpler solution though that would cover even more cases.

So basically we want to ensure the check is executed when user is logged in when:

* module === Widgetize
*  `[General] enable_framed_pages=1` or `[General]enable_framed_settings=1` config is enabled
* A third new check: We should also always execute the browser check if `Common::getRequestVar('token_auth', '', 'string')` is not empty. I think this would make both the previous two points obsolete as it should cover all the cases with widgetize and the 2 settings but just to be safe we may want to have this as yet another check :) 

And we always skip it when the API.index or API.glossary action is executed (see the code above).

Sorry for this being so complicated :) Happy to go over this tomorrow in the call :)

Reply:
Thanks, @tsteur for the detailed description, clear and easily understandable.
I make the change 👨‍💻

Reply:
@tsteur I updated the `FrontController.php`. 
Do you have any recommendations for tests? I made 4 UI tests. Any integration tests? Or more UI?
Thanks

Reply:
I think it be good to add a simple test for `checkIfBrowserSupported`. In this case I would have likely actually only created integration tests for `checkIfBrowserSupported ()` and for calling the `dispatch` method as integration or system tests are a lot faster to execute than UI tests and would have likely kind of achieved the same. Or one UI test and the rest integration/system tests. There isn't a need though to change this as everyone might do it differently and both solutions are legit. Generally, if we can test some behaviour using integration/system test like in this case, this would have probably been the preferred way because there isn't really any custom UI to test and it only checks if the exception is triggered. I will try to document this. It's hard to find the right balance there though and nobody gets it always right as it's not really clear when to do what.

Reply:
I created https://github.com/matomo-org/developer-documentation/commit/473259904455c866f7d8b6a8d468567e7660ec33#diff-6f870fb2ed7d589ffe2667ca2c4a9d2d3479fa9443a8a573d750bab93be0059f which should alter appear on https://developer.matomo.org/guides/tests-ui

Reply:
@tsteur just one more question about the tests.

I added tests for `checkIfBrowserSupported()` but the logic in the `FrontController` is not tested yet. Should we do something about that?

Reply:
The logic in the FrontController would be tested in the UI test so all good